storage/imx6ull-flash/flashdev.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Phoenix-RTOS
+ *
+ * IMX6ULL NAND flash device interface
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/minmax.h>
+
+#include "imx6ull-flashdev.h"
+#include "imx6ull-flashdrv.h"
+
+#define ECC_BITFLIP_THRESHOLD 10
+#define ECCN_BITFLIP_STRENGHT 14
+#define ECC0_BITFLIP_STRENGHT 16
+
+
+/* Auxiliary functions */
+
+/* Function derived from the previous code phoenix-rtos-filesystems/jffs/phoenix-rtos/mtd.c */
+static int _flashmtd_errRead(int ret, flashdrv_meta_t *meta, int status, void *data)
+{
+	int max_bitflip = 0;
+	unsigned *blk;
+	int i, j, err = 0;
+	/* check current read status. if it is EBADMSG we have nothing to do here since
+	 * read is marked as uncorrectable */
+	if (status == -EBADMSG) {
+		return status;
+	}
+
+	/* check status for every block */
+	for (i = 0; i < sizeof(meta->errors) / sizeof(meta->errors[0]); i++) {
+		/* no error or all ones */
+		if (meta->errors[i] == 0x0 || meta->errors[i] == 0xff) {
+			continue;
+		}
+		/* uncorrectable errors but maybe block is erased and we can still use it */
+		/* TODO: check if this actually works */
+		else if (meta->errors[i] == 0xfe) {
+			err = 0;
+			if (i == 0) {
+				blk = (unsigned *)meta->metadata;
+				for (j = 0; j < sizeof(meta->metadata) / sizeof(unsigned); ++j) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*blk + j);
+				}
+
+				if (err > ECC0_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, sizeof(meta->metadata));
+				max_bitflip = max(max_bitflip, err);
+			}
+			else {
+				blk = data + ((i - 1) * 512);
+				for (j = 0; j < 128; j++) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*(blk + j));
+				}
+
+				if (err > ECCN_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, 512);
+				max_bitflip = max(max_bitflip, err);
+			}
+		}
+		/* correctable errors */
+		else {
+			max_bitflip = max(max_bitflip, (int)meta->errors[i]);
+		}
+	}
+
+	return max_bitflip >= ECC_BITFLIP_THRESHOLD ? -EUCLEAN : status;
+}
+
+
+/* MTD interface */
+
+static int flashmtd_erase(struct _storage_t *strg, off_t offs, size_t size)
```

Comment:
I'd use `storage_t * strg` instead of `struct _storage_t *strg`. Repeats in other places in this PR.

Reply:
This change would require changes to the interface and dependent modules. I leave it as it is.